CNDB-11210: Fix false negatives for non-eq

k-rus opened the pull request:
Fixes correctness of executing queries with non-equality expressions
on columns in SAI, which can be truncated. Truncating can happen for
Decimal or Big integer types, which was leading to false negatives for
non-eq expressions, since it was negating the result of non-precise 
equality search in the index.

Fixes riptano/cnbd#11210

Comments to reviewers:
1. The added tests assume that the order in the query result is preserved as noticed in other tests. Let me know if it's incorrect assumption and needs to be fixed.
2. Let me know if I missed to use existing functions.
3. I might not be using idiomatic Java, thus let me know if there are places to improve.
4. The implementation keeps duplicated code, which can be refactored into a single function but seems to require introducing a help class. Let me know if I should do a better solution than now.

pkolaczk commented on src/java/org/apache/cassandra/db/marshal/AbstractType.java:
```diff
@@ -415,6 +415,11 @@ public boolean isTuple()
         return false;
     }
 
+    public boolean isTruncatable()
+    {
+        return false;
+    }
+
```
The truncation / rounding behavior is SAI specific. This shouldn't be in `db.marshal.AbstractType`, but rather in `TypeUtil` maybe?

k-rus replied:
Thank you for the suggestion. Actually there is already function doing what I need: `supportsRounding`. So I will just use it.

pkolaczk reviewed (changes requested):
Looks good. 
There are two minor things:
- the `isTruncatable` thing should be moved to SAI `TypeUtil`, as it is SAI specific. And `TypeUtil.encode` is responsible for all the truncation, so good to keep all related things together.
- you will need to update the formulas for NOT_xxx operators in the functions for estimating the number of rows matched in `SSTableIndex` and `MemtableIndex` after you rebase, so they reflect the new behavior - they are not present in the version of the code you worked on, because they were merged yesterday ;)

k-rus commented:
@pkolaczk I added Decimal type to `QueryWriteLifecycleTest`, so now the decimal cost model path is also covered. My understanding is that the test doesn't check correctness.

CI of my PR has some failures, but none of them seems to be related to the changes.

k-rus commented:
> @pkolaczk I added Decimal type to `QueryWriteLifecycleTest`, so now the decimal cost model path is also covered. My understanding is that the test doesn't check correctness.

I fixed the added test to run on different index formats in 51d69bda4877f493f078db1b7d741a7176e6e414
